File: Javascript/Add_Two_Numbers.js
// problem link: https://www.leetcode.com/problems/add-two-numbers/

// recursive call approach currently clocking @124ms runtime,
// faster than 90.92% amongst all other javascript submissions
// of the same problem.
const addTwoNumbers = (list1, list2, carry = 0) => {
    const sum = (list1?.val ?? 0) + (list2?.val ?? 0) + carry;
    carry = sum > 9 ? 1 : 0;
    if (
        (list1 === undefined || list1 === null || list1.next === null) &&
        (list2 === undefined || list2 === null || list2.next === null) &&
        carry == 0
    ) {
        return new ListNode(sum % 10, null);
    }
    return new ListNode(sum % 10, addTwoNumbers(list1?.next, list2?.next, carry));
};

File: GO/add_two_numbers.go
// Link: https://leetcode.com/problems/add-two-numbers/
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	sol := new(ListNode)
	var indexSol *ListNode
	pass1 := false

	for {
		sum := 0
		if l1 != nil || l2 != nil {
			if l1 != nil {
				sum = l1.Val
				l1 = l1.Next
			}
			if l2 != nil {
				sum += l2.Val
				l2 = l2.Next
			}
		} else if !pass1 {
			break
		}

		if pass1 {
			sum++
		}
		if sum > 9 {
			pass1 = true
		} else {
			pass1 = false
		}

		var val = sum % 10
		if indexSol == nil {
			sol.Val = val
			indexSol = sol
		} else {
			indexSol.Next = &ListNode{Val: val}
			indexSol = indexSol.Next
		}
	}

	return sol
}
